spoolman/database/spool.py

```diff
@@ -302,6 +324,68 @@ async def use_length(db: AsyncSession, spool_id: int, length: float) -> models.S
     return spool
 
 
+async def measure(db: AsyncSession, spool_id: int, weight: float) -> models.Spool:
+    """Record usage based on current gross weight of spool.
+
+    Increases the used_weight attribute of the spool.
+    Updates the first_used and last_used attributes where appropriate.
+
+    Args:
+        db (AsyncSession): Database session
+        spool_id (int): Spool ID
+        weight (float): Length of filament to consume, in mm
+
+    Returns:
+        models.Spool: Updated spool object
+
+    """
+    spool_result = await db.execute(
+        sqlalchemy.select(models.Spool.initial_weight, models.Spool.used_weight, models.Spool.empty_weight).where(
+            models.Spool.id == spool_id,
+        ),
+    )
+
+    try:
+        spool_info = spool_result.one()
+    except NoResultFound as exc:
+        raise SpoolMeasureError("Spool not found.") from exc
+
+    initial_weight = spool_info[0]
+    empty_weight = spool_info[2]
+    if initial_weight is None or empty_weight is None:
+        # Get filament weight and spool_weight
+        result = await db.execute(
+            sqlalchemy.select(models.Filament.weight, models.Filament.spool_weight)
+            .join(models.Spool, models.Spool.filament_id == models.Filament.id)
+            .where(models.Spool.id == spool_id),
+        )
+        try:
+            filament_info = result.one()
+        except NoResultFound as exc:
+            raise ItemNotFoundError("Filament not found for spool.") from exc
+        if empty_weight is None:
+            empty_weight = filament_info[1]
+
+        if initial_weight is None:
+            initial_weight = filament_info[0] + empty_weight
+
+    # Calculate the current gross weight (initial_weight - used_weight)
+    current_use = initial_weight - spool_info[1]
+
+    # if the measurement is greater than the initial weight, raise an error
+    if weight > current_use:
+        raise SpoolMeasureError("Measured weight cannot increase.")
```

Comment:
Maybe it should be able to? I don't see why not. Then you could occasionally re-measure your spools and if there has been some discrepancy such that Spoolman thinks more filament has been used than it actually has, it can be corrected.

Reply:
Right now this function utilizes the `use_weight` function, as to not replicate this functionality.  The original idea was that `measure` was just another flavor of `use` (like use weight or use length).  If I allow for the measured weight to effectively add weight to the spool, it should probably do so on it's own logical path.

If we want to allow for measurements to add, we can, but how would you want to go about this?  I'm not opposed, but I'd like to know how you would want that to look.

Reply:
use_weight supports negative numbers, so this restriction is completely arbitrary

Reply:
The issue I was running in to was what happens if your measurement is greater than the initial weight.  In this case, I just bump the initial weight to the measurement and set the used to 0.